src/fcast.c

```diff
@@ -15,57 +22,54 @@ SEXP fcast(SEXP lhs, SEXP val, SEXP nrowArg, SEXP ncolArg, SEXP idxArg, SEXP fil
     SET_VECTOR_ELT(ans, i, VECTOR_ELT(lhs, i));
   }
   // get val cols
+  bool some_fill = any_NA_int(nrows*ncols, idx);
```

Comment:
I understand that the alternate approach would be to suppress the warnings when we call `fun` on the empty vector to get the fill value. That is possible, but seems to me wasteful/un-necessary, because when there are no NA in idx, then the fill value is not used (so computing it is not necessary, and confusing, which was actually the core issue -- if a value is not needed, then why would we compute it?).

Reply:
Having a bit of trouble following the full context of Matt's comment here (probably elsewhere in the full comment history), but it looks similar to my reaction, which is some slight surprise that this is the approach for this functionality. I would have expected an approach that can tell on-the-fly (while the output is being constructed) whether the reshape is "complete" or "ragged" (the latter being where `fill=` is needed).

Is that not possible? If it's not possible, a comment here describing what was tried / why we think it can't be done would be great. If we're just settling on this approach for expedience and think it can be improved in the future, that may also be fine (pending performance check), but warrants a follow-up issue to revisit with instructions on what we're going for.

Reply:
the idxArg comes from this line of R code in fcast.R:
```r
    idx = do.call("CJ", mapunique)[map, 'I' := .I][["I"]] # TO DO: move this to C and avoid materialising the Cross Join.
```
so we could do anyNA(idx) and pass that through as another argument to fcast, but I suppose there would be no efficiency gain, and then we would have another argument to fcast, which seems desirable to avoid.

Reply:
If you think it's something worth pursuing in the longer term, please file a follow-up, but I'm happy for now.